maven-core/src/main/java/org/apache/maven/plugin/DefaultBuildPluginManager.java

```diff
@@ -143,7 +143,7 @@ public void executeMojo(MavenSession session, MojoExecution mojoExecution)
                 mojoExecutionListener.beforeMojoExecution(mojoExecutionEvent);
                 mojo.execute();
                 mojoExecutionListener.afterMojoExecutionSuccess(mojoExecutionEvent);
-            } catch (ClassCastException e) {
+            } catch (ClassCastException | MojoException e) {
```

Comment:
`MojoException` is actually a `RuntimeException`, so the outer catch block needs to be enhanced to rethrow directly instead of calling `mojoExecutionListener.afterExecutionFailure`.  Just to keep the old behaviour.

Reply:
Also, I would extend to any `MavenException`, since those are rather expected conditions.

Reply:
Done